### ISSUE TITLE
Allow to set 0 weight decay for biases and params in batch norm 

### ISSUE THREAD

**Issue by Jiaming-Liu** [pytorch/pytorch#1402]
This seems to be a good practice. But it is not realized in the optimizer class.

**Comment by Jiaming-Liu** [comment]
For this issue, I basically do something like this:
```python
def group_weight(module):
    group_decay = []
    group_no_decay = []
    for m in module.modules():
        if isinstance(m, nn.Linear):
            group_decay.append(m.weight)
            if m.bias is not None:
                group_no_decay.append(m.bias)
        elif isinstance(m, _ConvNd):
            group_decay.append(m.weight)
            if m.bias is not None:
                group_no_decay.append(m.bias)
        elif isinstance(m, _BatchNorm):
            if m.bias is not None:
                group_no_decay.append(m.weight)
            if m.bias is not None:
                group_no_decay.append(m.bias)

    assert len(list(module.parameters())) == len(group_decay) + len(group_no_decay)
    groups = [dict(params=group_decay), dict(params=group_no_decay, weight_decay=.0)]
    return groups
```

**Comment by fmassa** [comment]
Using the newly introduced `named_parameters` https://github.com/pytorch/pytorch/pull/1242 seems to be the easiest way of doing it I think

**Comment by soumith** [comment]
yes, with named_parameters it is very easy to filter all `bias` to separate per-parameter options. Maybe we can provide a helper function in utils.

**Comment by Jiaming-Liu** [comment]
Thank you for your reply. IMO this should be an option of either the __init__ of optimizers or the one of layers.

**Comment by soumith** [comment]
the optimizer itself is not aware of neural networks per se. So, it cant differentiate weight from bias.
Filtering the parameters into two sets via a one-line lambda is what we'd provide.

**Comment by turian** [comment]
Nonetheless, Facebook has an elegant method to `exclude_bias_and_norm` from `weight_decay` and `lars_adaptation` simply by checking if the parameter has `p.dim ==1`.

That is an agnostic approach and a decent option to add.

https://github.com/facebookresearch/vicreg/blob/main/main_vicreg.py#L235
